### ISSUE TITLE
Plugin upgrade breaks recognition of build tasks when using status bar button

### ISSUE THREAD

**Issue by jonbot1** [microsoft/vscode-cmake-tools#2959]
### Brief Issue Summary

With a default built task configured from task.json, pressing the Build button from the status bar now reveals a menu rather than running the default build task.  The menu does not appear to contain the default build task as an option. This was not a problem in 1.12.27 - it would launch the default task as expected.

Build button:
![image](https://user-images.githubusercontent.com/90410254/214101315-0adf57bc-b289-4881-a986-ecf6d1bd4ce7.png)

Revealed menu after pressing button:
![image](https://user-images.githubusercontent.com/90410254/214101581-d760033d-3d6a-406a-ba70-b13fb1ffe0af.png)

The "Cmake:Build Task" option in the workspace settings is set.

### CMake Tools Diagnostics

```shell
{
  "os": "linux",
  "vscodeVersion": "1.74.3",
  "cmtVersion": "1.13.40",
  "configurations": [
    {
      "folder": "/home/jon/git/proto",
      "cmakeVersion": "3.18.4",
      "configured": true,
      "generator": "Unix Makefiles",
      "usesPresets": false,
      "compilers": {
        "C": "/bin/x86_64-linux-gnu-gcc-10",
        "CXX": "/bin/x86_64-linux-gnu-g++-10"
      }
    }
  ],
  "cpptoolsIntegration": {
    "isReady": true,
    "hasCodeModel": true,
    "activeBuildType": "Debug",
    "buildTypesSeen": [
      "Debug"
    ],
    "requests": [],
    "responses": [],
    "partialMatches": [],
    "targetCount": 1,
    "executablesCount": 0,
    "librariesCount": 1,
    "targets": [
      {
        "name": "rpc",
        "type": "STATIC_LIBRARY"
      }
    ]
  },
  "settings": [
    {
      "communicationMode": "automatic",
      "useCMakePresets": "auto",
      "configureOnOpen": true
    }
  ]
}
```

### Debug Log

```shell
(no relevant input generated when the menu comes up)
```

### Additional Information

Below is the contents of my tasks.json.  Pressing ctrl+shift+b builds as expected.

```
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "label": "build current target",
      "type": "shell",
      "options": {
        "cwd": "${workspaceRoot}/build"
      },
      "command": "make -j`nproc` ${command:cmake.buildTargetName}",
      "problemMatcher": [],
      "group": {
        "kind": "build",
        "isDefault": true
      }
    },
  ]
}
```
Setting the default built task again (ctrl+shift+p -> Configure Default Build Task) does not resolve the problem.

**Comment by elahehrashedi** [comment]
Looking closer at your task's definition, I noticed that your task is a `"shell" `type not a `"cmake" `type. 
In this design, for a task to be used as a build task by CMake, it needs to be a "cmake" type. 
Can you try this task instead and let us know if it works for you?
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "build current target",
      "type": "cmake",
      "command": "build",
      "targets": ["all"]
      "problemMatcher": [],
      "group": {
        "kind": "build",
        "isDefault": true
      }
    },
  ]
}
```
and set the build args in settings.json as below:
```json
     "cmake.buildDirectory": "${workspaceFolder}/build",
    "cmake.parallelJobs": "9",
```

**Comment by jonbot1** [comment]
> Looking closer at your task's definition, I noticed that your task is a `"shell" `type not a `"cmake" `type. In this design, for a task to be used as a build task by CMake, it needs to be a "cmake" type. Can you try this task instead and let us know if it works for you?
> 
> ```json
> {
>   "version": "2.0.0",
>   "tasks": [
>     {
>       "label": "build current target",
>       "type": "cmake",
>       "command": "build",
>       "targets": ["all"]
>       "problemMatcher": [],
>       "group": {
>         "kind": "build",
>         "isDefault": true
>       }
>     },
>   ]
> }
> ```
> 
> and set the build args in settings.json as below:
> 
> ```json
>      "cmake.buildDirectory": "${workspaceFolder}/build",
>     "cmake.parallelJobs": "9",
> ```

Hi,

Thanks for the quick reply.  That does work around the problem, but it undermines our purpose in having a task-based build. Using "shell" was intentional because we get colored text in our output window (which can be helpful for a large and complex build), we can query the processor count rather than hardcoding it into the settings.json, and just generally we have the option of incorporating other arbitrary build enhancements. 

In this way, the task you provided gets past the bug but it doesn't convey much benefit beyond just doing a non-task-based build.

I can't speak to the broader intent of whether 'shell' tasks were originally intended to work, but it does work in the last version, and it's hard to see how a task-based build is useful if you can't customize how the build is executed more than with a non-task-based build.

**Comment by jonbot1** [comment]
Yes - the intent is to use the build button on the status bar, as that's right next to where you set the target you're building.  It worked before the update.

**Comment by piomis** [comment]
@jonbot1 it wrked before update but it worked 'to well' - target set on status bar was used by any build task. So even if you select other target from 'Cmake Project Outline' you Got only build of one set in status bar.
From that reason searching of tasks had to be redesigned and plugin itself assumed to use only 'cmake' tasks. Of course it is still possible to run non-cmake tasks with 'Run Task' command.

**Comment by jonbot1** [comment]
Thanks @piomis.  Can you explain then what the status bar selector is now for if not to choose the target to build?  I would think an implementation where choosing an item in the Cmake Project Outline updates the status bar selector would be reasonable...but I'm not so well-versed in the tool to claim I know all the knock-on effects of that.

Is there a way to define a custom build command for a "cmake" task type?

**Comment by piomis** [comment]
@jonbot1
First of all small disclaimer.
I am not an owner of cmake-tools thus I may only answer what I think or believe - please do not take it as official statement from the team.

In regards to first question:
Target selected in status bar (as also hoover tool tip says) is "default target" which is used by default if none of other targets is used.
Thus changing it when specific target is selected from "Cmake Project outline" (or task with target already specified in tasks.json) wouldn't make much sense.

In regards to custom build command. I am not sure what "custom command" you mean.
For any specific custom stuff to do I personally use 'cmake' capabilities itself (e.g. `add_custom_target`)
If you need to specify number of "parallel jobs" I believe cmake-tools already supports automatic detection  - see `cmake.parallelJobs`
If you need some additional options to be passed I suppose you may try "options" in task definition in tasks.json

**Comment by 2997215859** [comment]
I encountered the same problem, and resolved it by returning to the old version 1.27.*

**Comment by jonbot1** [comment]
"Target selected in status bar (as also hoover tool tip says) is "default target" which is used by default if none of other targets is used.
Thus changing it when specific target is selected from "Cmake Project outline" (or task with target already specified in tasks.json) wouldn't make much sense."

To me this defies sensible usability. The setting on the status bar is the most visible setting - it's always there, taking up real-estates. To saying that you set the default there but have to dig into a menu/view to set what overrides it seems fundamentally backwards to me. I've never used the CMake project outline, and with a clean button that displays a searchable pull-down of targets, I don't know why I ever would.

It's possible that I can implement particular behaviors using cmake, but in the case of rich terminal output (errors getting highlighted in a different color, for example), it's not clear to me that would be possible...

I understand that perhaps isn't your call, @piomis , and also I probably don't understand all the use cases. I'm going to step back from this but it doesn't seem like the new way of working makes much sense based on my own perspective.  I'll be trying to avoid upgrading the plugin as long as possible as a workaround.

**Comment by piomis** [comment]
CMake Outline lists all targets defined (AFAIR CMake GUI provids similar capability).
For me it is very useful for so called 'utility' targets (these are onea which are not 'real' build but helpers/additional activities). 
I usually use status bar target is for switching between different 'build' targets (e.g. Application1, library, etc) which are triggered by 'build' button.
But utility or test tasks IMHO do not fit into that workflow.

Nevertheless I understand your concern that 'shell' task type stopped working. I am also not fully satisfied with current solution. I will try finding some time and propose pull request and lets see what maintainers will say.

**Comment by gcampbell-msft** [comment]
@jonbot1 @piomis It seems based on all of the conversation above that you all have diagnosed what is happening. We are depending on a task being of type "cmake" in order for it to be used from the "Build" button. 

Therefore, this isn't as much of a bug, as it is currently by design. 

However, it's a very reasonable request, so I am switching it from a "bug" to a "Feature Request" and moving it to the 1.16 milestone.
